### ISSUE TITLE
Building parsers programmatically (question)

### ISSUE THREAD

**Issue by catie** [keichi/binary-parser#237]
I'm new to Node, so please forgive me if I'm missing something obvious here. I'm building a utility to parse binary files based on a Json definition of the file specification, and while I originally implemented it in Java, that solution is clunky and already showing strain so I'm rebuilding in Node.

My Json elements have types that already match your primitives, but I notice the `Parser.primitiveN()` function is private. It seems a little silly to re-implement mappings that I can see are happening in reverse once I call `parse()`, but I could do that if needed. I'm also happy to make the changes to expose `PrimitiveTypes` and `primitiveN()`, but looking at the git history it seems like they might've been made private for a reason. If someone could shed some light on why and provide guidance on the best path forward it would be greatly appreciated. Thanks!

**Comment by keichi** [comment]
Sorry for not getting back to you sooner. IIRC `primitiveN` is private simply because there were no use cases to generate the parser definition itself dynamically. Also we wanted to minimize the API surface.

Feel free to open a PR to make it public!

**Comment by catie** [comment]
Makes sense! I'll try and find some time to clean up my fork and get it submitted. Thanks @keichi!

**Comment by keichi** [comment]
You can build parsers programmatically by calling the public methods dynamically using bracket notation:

```js
const parser = new Parser();
const typeMap = { "uint8": "uint8", "uint16": "uint16be", ... };

fields.forEach(field => {
  parser[typeMap[field.type]](field.name, field.options);
});
```

While `primitiveN()` is private, the public methods (`uint8()`, `uint16be()`, etc.) are fully accessible and can be called dynamically. Closing as answered.
